File: lib/model/elo.ts
import { ELO_SEED_RATING, K_FACTORS } from "@/lib/constants";

/** Win expectancy for a side with net rating advantage `ratingDiff` (incl. HA). */
export function expectedScore(ratingDiff: number): number {
  return 1 / (1 + Math.pow(10, -ratingDiff / 400));
}

/** World Football Elo goal-difference multiplier. */
export function gFactor(goalDiff: number): number {
  const n = Math.abs(goalDiff);
  if (n <= 1) return 1;
  if (n === 2) return 1.5;
  return (11 + n) / 8;
}

/** Map a martj42 tournament label to a World Football Elo K-factor. */
export function kFactorForTournament(tournament: string): number {
  const t = tournament.toLowerCase();
  if (t.includes("qualification")) return K_FACTORS.qualifier;
  if (t === "fifa world cup") return K_FACTORS.worldCup;
  if (t === "friendly") return K_FACTORS.friendly;
  const continental = [
    "copa america",
    "uefa euro",
    "african cup of nations",
    "africa cup of nations",
    "afc asian cup",
    "gold cup",
    "concacaf championship",
    "oceania nations cup",
    "ofc nations cup",
    "confederations cup",
  ];
  const plain = t
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "");
  if (continental.some((c) => plain.includes(c))) return K_FACTORS.continentalFinal;
  return K_FACTORS.tournament;
}

/**
 * Elo update for one match. Returns the rating deltas (home delta = -away delta).
 * `homeAdvantage` must be 0 for neutral-venue matches.
 */
export function eloDelta(
  ratingHome: number,
  ratingAway: number,
  homeGoals: number,
  awayGoals: number,
  k: number,
  homeAdvantage: number
): number {
  const we = expectedScore(ratingHome + homeAdvantage - ratingAway);
  const w = homeGoals > awayGoals ? 1 : homeGoals === awayGoals ? 0.5 : 0;
  const g = gFactor(homeGoals - awayGoals);
  return k * g * (w - we);
}

export interface HistoricalMatch {
  date: string;
  home: string;
  away: string;
  homeScore: number;
  awayScore: number;
  tournament: string;
  neutral: boolean;
}

export interface HistoryEntry {
  rating: number;
  matches: number;
  lastMatch: string;
}

/**
 * Run Elo over a chronologically sorted match history, keyed by raw team
 * names (canonical mapping happens after, for the 48 WC teams only).
 */
export function runHistory(
  matches: HistoricalMatch[],
  homeAdvantage: number
): Map<string, HistoryEntry> {
  const table = new Map<string, HistoryEntry>();
  const get = (team: string): HistoryEntry => {
    let e = table.get(team);
    if (!e) {
      e = { rating: ELO_SEED_RATING, matches: 0, lastMatch: "" };
      table.set(team, e);
    }
    return e;
  };
  for (const m of matches) {
    const home = get(m.home);
    const away = get(m.away);
    const k = kFactorForTournament(m.tournament);
    const delta = eloDelta(
      home.rating,
      away.rating,
      m.homeScore,
      m.awayScore,
      k,
      m.neutral ? 0 : homeAdvantage
    );
    home.rating += delta;
    away.rating -= delta;
    home.matches++;
    away.matches++;
    home.lastMatch = m.date;
    away.lastMatch = m.date;
  }
  return table;
}

File: ml/elo.py
"""Running World Football Elo over the full match history.

Produces a pre-match Elo for both teams of every match (the time-varying
strength feature the GBM trains on) plus the final current ratings used to seed
the 2026 export. Mirrors lib/model/elo.ts so TS and Python agree.
"""
from __future__ import annotations

import unicodedata
from dataclasses import dataclass

import numpy as np
import pandas as pd


@dataclass
class EloConfig:
    k_world_cup: int = 60
    k_continental_final: int = 50
    k_qualifier: int = 40
    k_tournament: int = 30
    k_friendly: int = 20
    home_advantage: int = 80
    seed_rating: int = 1500


_CONTINENTAL = (
    "copa america", "uefa euro", "african cup of nations", "africa cup of nations",
    "afc asian cup", "gold cup", "concacaf championship", "oceania nations cup",
    "ofc nations cup", "confederations cup",
)


def _strip(s: str) -> str:
    s = unicodedata.normalize("NFD", s)
    return "".join(c for c in s if unicodedata.category(c) != "Mn").lower()


def k_factor(tournament: str, cfg: EloConfig) -> int:
    t = str(tournament).lower()
    if "qualification" in t:
        return cfg.k_qualifier
    if t == "fifa world cup":
        return cfg.k_world_cup
    if t == "friendly":
        return cfg.k_friendly
    plain = _strip(t)
    if any(c in plain for c in _CONTINENTAL):
        return cfg.k_continental_final
    return cfg.k_tournament


def expected_score(rating_diff: float) -> float:
    return 1.0 / (1.0 + 10.0 ** (-rating_diff / 400.0))


def g_factor(goal_diff: int) -> float:
    n = abs(int(goal_diff))
    if n <= 1:
        return 1.0
    if n == 2:
        return 1.5
    return (11 + n) / 8.0


def run_elo(matches: pd.DataFrame, cfg: EloConfig) -> pd.DataFrame:
    """Annotate each row with pre-match elo_home / elo_away.

    `matches` must be sorted by date with columns: home_team, away_team,
    home_score, away_score, tournament, neutral (bool).
    Returns the same frame with elo_home, elo_away columns added, plus the
    final ratings accessible via the returned frame's attrs['final_ratings'].
    """
    ratings: dict[str, float] = {}

    def get(team: str) -> float:
        return ratings.get(team, float(cfg.seed_rating))

    elo_home = np.empty(len(matches))
    elo_away = np.empty(len(matches))

    for i, (_, m) in enumerate(matches.iterrows()):
        h, a = m["home_team"], m["away_team"]
        rh, ra = get(h), get(a)
        elo_home[i] = rh
        elo_away[i] = ra
        ha = 0 if m["neutral"] else cfg.home_advantage
        we = expected_score(rh + ha - ra)
        hs, as_ = int(m["home_score"]), int(m["away_score"])
        w = 1.0 if hs > as_ else 0.5 if hs == as_ else 0.0
        k = k_factor(m["tournament"], cfg)
        delta = k * g_factor(hs - as_) * (w - we)
        ratings[h] = rh + delta
        ratings[a] = ra - delta

    out = matches.copy()
    out["elo_home"] = elo_home
    out["elo_away"] = elo_away
    out.attrs["final_ratings"] = dict(ratings)
    return out
